### ISSUE TITLE
while using `mimalloc` memory allocator, rules are mistaken with regex-based.

### ISSUE THREAD

**Issue by enihcam** [miniupnp/miniupnp#652]
```
miniupnpd.conf starting from line 184:
---
allow 1024-65535 192.168.0.0/24 1024-65535
deny 0-65535 0.0.0.0/0 0-65535
```

```
error:
---
Apr 05 03:32:49 miniupnpd[423]: MiniUPnP is not compiled with ENABLE_REGEX. Please remove any regex filter and restart.
Apr 05 03:32:49 miniupnpd[423]: parsing error file /etc/miniupnpd/miniupnpd.conf line 184 : allow 1024
Apr 05 03:32:49 miniupnpd[423]: Error reading configuration file /etc/miniupnpd/miniupnpd.conf
```

is the following line related?
https://github.com/miniupnp/miniupnp/blob/2eecb06eb0c78a0d2acc88e920ca446d86a486e5/miniupnpd/upnppermissions.c#L333

**Comment by enihcam** [comment]
@epoupon

**Comment by miniupnp** [comment]
looks like a bug indeed

**Comment by miniupnp** [comment]
@enihcam could you try with this patch 5ca1a82 from branch [652-fix](https://github.com/miniupnp/miniupnp/tree/652-fix) ?

**Comment by enihcam** [comment]
> @enihcam could you try with this patch [5ca1a82](https://github.com/miniupnp/miniupnp/commit/5ca1a82f7fa9720b0be0fa3e2e7f4612b7192b33) from branch [652-fix](https://github.com/miniupnp/miniupnp/tree/652-fix) ?

i'm out of town for a couple of days. will get back to you asap.

**Comment by enihcam** [comment]
> @enihcam could you try with this patch [5ca1a82](https://github.com/miniupnp/miniupnp/commit/5ca1a82f7fa9720b0be0fa3e2e7f4612b7192b33) from branch [652-fix](https://github.com/miniupnp/miniupnp/tree/652-fix) ?

unfortunately the patch does not work. same error.

@epoupon from mimalloc project maybe can take a look?

**Comment by epoupon** [comment]
Sorry I am not from mimalloc project, just a mimalloc integrator/user.
The problem I reported is not likely to be yours, as it only affected non standard plateforms and had impacts on internal mimalloc's control structures.
Maybe you can try to repro with valgrind to help you spot uninitialized memory operations.

**Comment by enihcam** [comment]
@daanx from mimalloc project maybe can take a look?

@epoupon sorry, looped a wrong guy. :)

**Comment by enihcam** [comment]
@miniupnp is it possible to add more check conditions to workaround this issue? thanks.

**Comment by enihcam** [comment]
here's the root cause analysis from author of `mimalloc`:

> After more debugging, it seems miniupnp does a `token = realloc(token,i)` where `i` is zero .. (`upnppermissions.c:155`). Then, later, the pointer is dereferenced and tested (`perm->re[0] == '\0'`). Now, the C standard for a zero reallocation says:
> 
> _If the size of the space requested is zero, the behavior is implementation-defined: either a null pointer is returned, or the behavior is as if the size were some nonzero value, except that the returned pointer shall not be used to access an object_ ([link](https://stackoverflow.com/questions/16759849/using-realloc-x-0-instead-of-free-and-using-malloc-with-length-of-a-string)) and I think in C++ it is even undefined.
> 
> So, I think this is a bug in miniupnpd. Mimalloc always returns a pointer on a zero-sized (re)allocation. This way, realloc(p,0) always only returns NULL on an error where `p` is not freed.
> 


@miniupnp could you help address this issue? thanks a lot!

**Comment by miniupnp** [comment]
@enihcam please test the branch [652-fix](https://github.com/miniupnp/miniupnp/tree/652-fix), commit 861298fa2467c3bf0c2893b2b675aa2ab836e0a2

**Comment by enihcam** [comment]
@miniupnp your patch worked! thanks for quick response. :)

```bash
/data/miniupnp/miniupnpd $ ./testupnppermissions "allow 1024-65535 192.168.0.0/24 1024-65535"
 1 'allow 1024-65535 192.168.0.0/24 1024-65535'
Permission read successfully
allow 1024-65535 192.168.0.0/255.255.255.0 1024-65535

/data/miniupnp/miniupnpd $ LD_PRELOAD=../../mimalloc/out/debug/libmimalloc-debug.so ./testupnppermissions "allow 1024-65535 192.168.0.0/24 1024-65535"
 1 'allow 1024-65535 192.168.0.0/24 1024-65535'
Permission read successfully
allow 1024-65535 192.168.0.0/255.255.255.0 1024-65535
```
